File: comic/src/main/java/com/song/sunset/comic/widget/SunsetWidget.java
package com.song.sunset.comic.widget;

import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.Context;
import android.content.Intent;
import android.util.Log;
import android.widget.RemoteViews;

import com.song.sunset.comic.ComicCollectionActivity;
import com.song.sunset.comic.R;

import java.util.Arrays;

/**
 * Implementation of App Widget functionality.
 */
public class SunsetWidget extends AppWidgetProvider {

    public static final String FROM_WIDGET = "com.song.sunset.from_widget";

    static void updateAppWidget(Context context, AppWidgetManager appWidgetManager,
                                int appWidgetId) {

        // Construct the RemoteViews object
        RemoteViews views = new RemoteViews(context.getPackageName(), R.layout.sunset_widget);

        Intent intent = new Intent();
        intent.setClassName(context, "com.song.sunset.activitys.MainActivity");

        PendingIntent newestComic = PendingIntent.getActivity(context, (int) (Math.random() * 100000),
                intent, PendingIntent.FLAG_UPDATE_CURRENT);
        views.setOnClickPendingIntent(R.id.appwidget_newest_comic, newestComic);

        Intent collectedComicIntent = new Intent(context, ComicCollectionActivity.class);
        collectedComicIntent.putExtra(FROM_WIDGET, true);
        PendingIntent collectedComic = PendingIntent.getActivity(context, (int) (Math.random() * 100000),
                collectedComicIntent, PendingIntent.FLAG_UPDATE_CURRENT);
        views.setOnClickPendingIntent(R.id.appwidget_collected_comic, collectedComic);

        PendingIntent news = PendingIntent.getActivity(context, (int) (Math.random() * 100000),
                intent, PendingIntent.FLAG_UPDATE_CURRENT);
        views.setOnClickPendingIntent(R.id.appwidget_news, news);

        Intent newIntent = new Intent();
        newIntent.setClassName(context, "com.song.sunset.activitys.VideoListActivity");
        newIntent.putExtra(FROM_WIDGET, true);
        PendingIntent newsVideo = PendingIntent.getActivity(context, (int) (Math.random() * 100000),
                newIntent, PendingIntent.FLAG_UPDATE_CURRENT);
        views.setOnClickPendingIntent(R.id.appwidget_news_video, newsVideo);

        // Instruct the widget manager to update the widget
        appWidgetManager.updateAppWidget(appWidgetId, views);
    }

    @Override
    public void onReceive(Context context, Intent intent) {
        super.onReceive(context, intent);
    }

    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
        // There may be multiple widgets active, so update all of them
        Log.i("SunsetWidget", Arrays.toString(appWidgetIds));
        for (int appWidgetId : appWidgetIds) {
            updateAppWidget(context, appWidgetManager, appWidgetId);
        }
    }

    @Override
    public void onEnabled(Context context) {
        // Enter relevant functionality for when the first widget is created
    }

    @Override
    public void onDisabled(Context context) {
        // Enter relevant functionality for when the last widget is disabled
    }
}

File: comic/src/main/java/com/song/sunset/comic/widget/appwidget/SunsetWidget.kt
package com.song.sunset.comic.widget.appwidget

import android.appwidget.AppWidgetProvider
import android.content.Intent
import android.appwidget.AppWidgetManager
import android.widget.RemoteViews
import com.song.sunset.comic.R
import android.app.PendingIntent
import android.content.Context
import android.util.Log
import com.song.sunset.comic.ComicCollectionActivity

/**
 * Implementation of App Widget functionality.
 */
class SunsetWidget : AppWidgetProvider() {
    override fun onReceive(context: Context, intent: Intent) {
        super.onReceive(context, intent)
    }

    override fun onUpdate(context: Context, appWidgetManager: AppWidgetManager, appWidgetIds: IntArray) {
        // There may be multiple widgets active, so update all of them
        Log.i("SunsetWidget", appWidgetIds.contentToString())
        for (appWidgetId in appWidgetIds) {
            updateAppWidget(context, appWidgetManager, appWidgetId)
        }
    }

    override fun onEnabled(context: Context) {
        // Enter relevant functionality for when the first widget is created
    }

    override fun onDisabled(context: Context) {
        // Enter relevant functionality for when the last widget is disabled
    }

    companion object {
        const val FROM_WIDGET = "com.song.sunset.from_widget"
        fun updateAppWidget(context: Context, appWidgetManager: AppWidgetManager,
                            appWidgetId: Int) {

            //图标一
            val views = RemoteViews(context.packageName, R.layout.sunset_widget)
            val intent = Intent()
            intent.setClassName(context, "com.song.sunset.activitys.MainActivity")
            val newestComic = PendingIntent.getActivity(context, (Math.random() * 100000).toInt(),
                    intent, PendingIntent.FLAG_UPDATE_CURRENT)
            views.setOnClickPendingIntent(R.id.appwidget_newest_comic, newestComic)

            //图标二
            val pendingIntent = Intent(context, ComicCollectionActivity::class.java)
                    .let {
                        PendingIntent.getActivity(context, 0, it, PendingIntent.FLAG_UPDATE_CURRENT)
                    }
            views.setOnClickPendingIntent(R.id.appwidget_collected_comic, pendingIntent)

            //图标三
            val news = PendingIntent.getActivity(context, (Math.random() * 100000).toInt(),
                    intent, PendingIntent.FLAG_UPDATE_CURRENT)
            views.setOnClickPendingIntent(R.id.appwidget_news, news)

            //图标四
            val videoPendingIntent = Intent().apply {
                setClassName(context, "com.song.sunset.activitys.VideoListActivity")
                putExtra(FROM_WIDGET, true)
            }.let {
                PendingIntent.getActivity(context, (Math.random() * 100000).toInt(),
                        it, PendingIntent.FLAG_UPDATE_CURRENT)
            }
            views.setOnClickPendingIntent(R.id.appwidget_news_video, videoPendingIntent)

            appWidgetManager.updateAppWidget(appWidgetId, views)
        }
    }
}